inc/fftw.hpp

```diff
@@ -0,0 +1,629 @@
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // RUNTIME TIME FIELDS
+    
+      size_t n_;        // =[1]*..*[dim]
+      size_t n_padded_; // =[1]*..*[dim-1]*([dim]/2+1)
+      Extent extents_;
+
+      PlanType			plan_           ;
+      value_type*		data_           = nullptr;
+      ComplexType*       	data_transform_ = nullptr; // intermediate buffer
+      size_t             	data_size_;
+      size_t             	data_transform_size_;
+
+      FftwImpl(const Extent& cextents)
+	: n_(0),
+	  n_padded_(0),
+	  extents_(),
+	  plan_(),
+	  data_(nullptr),
+	  data_transform_(nullptr),
+	  data_size_(0),
+	  data_transform_size_(0)
+      {
+	extents_ = interpret_as::column_major(cextents);
+	
+	n_ = std::accumulate(extents_.begin(),
+			     extents_.end(),
+			     1,
+			     std::multiplies<unsigned>());
+	if(Padding)
+	  n_padded_ = n_ / extents_[NDim-1] * (extents_[NDim-1]/2 + 1);
+
+	data_size_ = ( Padding ? 2*n_padded_ : n_ ) * sizeof(value_type);
+	data_transform_size_ = IsInplace ? 0 : n_ * sizeof(ComplexType);
+
+	traits::thread_api<TPrecision>::init_threads();
+	traits::thread_api<TPrecision>::plan_with_threads(-1);
+      }
+
+      ~FftwImpl(){
+	traits::thread_api<TPrecision>::cleanup_threads();
+      }
+      
+      /**
+       * Returns allocated memory for FFT
+       */
+      size_t getAllocSize() {
+	return data_size_ + data_transform_size_;
+      }
+
+      // create FFT plan handle
+      void init_forward() {
+
+	//NOTE: contents of data_ et al are overwritten by this call
+	plan_ = traits::plan<TPrecision>::create(extents_, data_, data_transform_, traits::forward);
+      }
+
+      // recreates plan if needed
+      void init_backward() {
+
+	traits::plan<TPrecision>::destroy(plan_);
+
+	//NOTE: contents of data_ et al are overwritten by this call
+	plan_ = traits::plan<TPrecision>::create(extents_, data_, data_transform_, traits::backward);
+      
+      }
+
+      /**
+       * Returns estimated allocated memory on device for FFT plan, accessing the plan size in fftw is impossible
+       
+       */
+      size_t getPlanSize() {
+      
+	if(IsInplace)
+	  return data_size_;
+	else
+	  return data_size_ + data_transform_size_;
+	
+      }
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // --- next methods are benchmarked ---
+
+      void malloc() {
+
+	data_ = (value_type*)traits::memory_api<TPrecision>::malloc(data_size_);
+      
+	if(IsInplace){
+	  data_transform_ = reinterpret_cast<ComplexType*>(data_);
+	}
+	else{
+	  data_transform_ = (ComplexType*)traits::memory_api<TPrecision>::malloc(data_transform_size_);
+	}
+		   
+      }
+
+
+      void execute_forward() {
+	traits::plan<TPrecision>::execute(plan_);
+      }
+
+      void execute_backward() {
+	traits::plan<TPrecision>::execute(plan_);
+      }
+
+      template<typename THostData>
+      void upload(THostData* input) {
+
```

Comment:
input contains the example data, which should be loaded into the fftw owned array.
input probably could be reused, then copy in download() is not necessary.
In the other case we could benchmark fftw malloc. So, we have two benchmark cases for fftw now.
I would prefer the own fftw malloc'd array (might be proper aligned for SSE/AVX), and in the statistics we could separate the timings, what do you think?